feat: `rover template`

EverlastingBugstopper opened the pull request:
this PR adds `rover new` which downloads and unpacks GraphQL projects into a local directory.

EverlastingBugstopper committed:
wip: initial commit for rover subgraph init

EverlastingBugstopper committed:
chore: removes autogenerated .apollo

EverlastingBugstopper committed:
feat: improves info for filesystem read/write

EverlastingBugstopper committed:
use saucer

EverlastingBugstopper committed:
feat: minimal rover subgraph init

EverlastingBugstopper committed:
feat: read from .apollo/config.yaml for subgraph publish

EverlastingBugstopper committed:
feat: initial rover dev

EverlastingBugstopper committed:
feat: subgraph runner

EverlastingBugstopper committed:
properly join command runners

EverlastingBugstopper committed:
correct dependency

EverlastingBugstopper committed:
git saucer

EverlastingBugstopper committed:
wip

EverlastingBugstopper committed:
Merge branch 'main' into avery/subgraph-init

EverlastingBugstopper committed:
Merge branch 'avery/subgraph-init' into avery/alt-subgraph-dev

EverlastingBugstopper committed:
feat: extend other rover dev processes, install router

EverlastingBugstopper committed:
wait on router in separate thread

EverlastingBugstopper committed:
adds example

EverlastingBugstopper committed:
try to repro w/manual compose invocations

EverlastingBugstopper committed:
chore: adds --debug-socket option to send arbitrary messages to the supergraph socket

EverlastingBugstopper committed:
Merge branch 'main' into avery/subgraph-init

EverlastingBugstopper committed:
Merge branch 'avery/subgraph-init' into avery/alt-subgraph-dev

EverlastingBugstopper committed:
make it work on mac kinda

EverlastingBugstopper committed:
i think it works now

EverlastingBugstopper committed:
allow specifying a running graphql server

EverlastingBugstopper committed:
fix file watching on linux

EverlastingBugstopper committed:
fix: make it work on windows

EverlastingBugstopper committed:
better command cleanup

EverlastingBugstopper committed:
no dot-apollo for now

EverlastingBugstopper committed:
removes dot-apollo code, refactors runners

EverlastingBugstopper committed:
removes unused opts

EverlastingBugstopper committed:
removes unused opts

EverlastingBugstopper committed:
Merge branch 'main' into avery/alt-subgraph-dev

EverlastingBugstopper committed:
refactor

EverlastingBugstopper committed:
add/remove/update subgraphs w/live reload

EverlastingBugstopper committed:
clean up logging and endpoint detection

EverlastingBugstopper committed:
Merge branch 'main' into avery/alt-subgraph-dev

EverlastingBugstopper committed:
fixup introspection watching

EverlastingBugstopper committed:
fixup: removes error message passer

EverlastingBugstopper committed:
fixup macos, introspector, and add supergraph-demo nodemon

EverlastingBugstopper committed:
Merge branch 'main' into avery/alt-subgraph-dev

EverlastingBugstopper committed:
add back retries for select cases

EverlastingBugstopper committed:
test go green

EverlastingBugstopper committed:
remove tarball

EverlastingBugstopper committed:
dont change to localhost

EverlastingBugstopper committed:
add readme to supergraph-demo and use SocketAddr

EverlastingBugstopper committed:
adds prerequisites to demo

EverlastingBugstopper committed:
add npm run dev scripts

EverlastingBugstopper committed:
dont use this in prod

EverlastingBugstopper committed:
pass through subgraph errors

EverlastingBugstopper committed:
allow resolving pandas

EverlastingBugstopper committed:
chore: rayonize endpoint detection

EverlastingBugstopper committed:
more durable port detection/subgraph collision errors

EverlastingBugstopper committed:
fixup file watcher

EverlastingBugstopper committed:
fixup filewatcher

EverlastingBugstopper committed:
chore: clean up introspection example

EverlastingBugstopper committed:
Merge branch 'main' into avery/alt-subgraph-dev

EverlastingBugstopper committed:
address lints

EverlastingBugstopper committed:
fix: wait for socket to bind before starting supergraph

EverlastingBugstopper committed:
chore: clean up introspection errors and success

EverlastingBugstopper committed:
docs: updates help text and refactors subgraph name prompt (#1218)

Co-authored-by: Stephen Barlow <stephen@apollographql.com>

EverlastingBugstopper committed:
fix(dev): reduce introspection timeouts + spawn/kill process _groups_ (#1233)

EverlastingBugstopper committed:
fix: prompt for schema when it is detected (#1236)

EverlastingBugstopper committed:
fixup: default yes confirm works now

EverlastingBugstopper committed:
release: v0.9.0-alpha.0

EverlastingBugstopper committed:
fixup: rayon spawn introspectors (#1238)

EverlastingBugstopper committed:
release: v0.9.0-alpha.1

EverlastingBugstopper committed:
chore: removes --command and port scanning (#1240)

EverlastingBugstopper committed:
release: v0.9.0-alpha.2

EverlastingBugstopper committed:
Merge branch 'main' into avery/alt-subgraph-dev

EverlastingBugstopper committed:
fix: fixes up socket message passing (#1252)

EverlastingBugstopper committed:
release: v0.9.0-alpha.3

EverlastingBugstopper committed:
release: v0.9.0-alpha.4

EverlastingBugstopper committed:
chore: removes superfluous output

EverlastingBugstopper committed:
fix: kill all sessions when main session shuts down

EverlastingBugstopper committed:
release: v0.9.0-alpha.5

EverlastingBugstopper committed:
Merge branch 'main' into avery/alt-subgraph-dev

abernix committed:
enable exposing the query plan within `rover dev`

Implements the experimental expose query plan feature in `rover dev` until
we enable Apollo Router to negotiate this itself via an env variable.

Past Work: https://github.com/apollographql/router/issues/1075
Future Work: https://github.com/apollographql/router/issues/1474

EverlastingBugstopper committed:
Merge branch 'main' into avery/alt-subgraph-dev

EverlastingBugstopper committed:
feat: allow configurable router downloads

EverlastingBugstopper committed:
feat: adds `--port` argument to `rover dev` (#1275)

fixes #1231 and #1266 by adding a `--port` to `rover dev` that configures the port that the router listens on.

each unique port creates its own `rover dev` session that will not interfere with routers running on other ports. additionally, if you try to run a subgraph on the same port as the router, you will get an error message letting you know, rather than the previous confusing message "there is a subgraph already running on that port".

EverlastingBugstopper committed:
chore: refactors elv2 prompts for rover dev and rover supergraph compose (#1276)

fixes #1265 by requiring users to accept the ELv2 license immediately upon command invocation rather than only on the plugin installation itself.

EverlastingBugstopper committed:
fixup help text

EverlastingBugstopper committed:
release: v0.9.0-alpha.6

EverlastingBugstopper committed:
feat: composition errors are printed in the session that caused the error (#1282)

fixes #1257 and #1277 by upgrading the message protocol to send composition errors back and forth between the follower and the leader sessions. 

composition errors no longer cause crashes, and they are reported in the session that caused the error, in addition to the main session.

EverlastingBugstopper committed:
release: v0.9.0-alpha.7

EverlastingBugstopper committed:
Merge branch 'main' into avery/alt-subgraph-dev

michael-watson committed:
Initial MVP

Still needs orbit

EverlastingBugstopper commented on Cargo.toml:
```diff
@@ -68,12 +68,14 @@ ansi_term = "0.12"
 atty = "0.2"
 buildstructor = "0.3"
 calm_io = "0.1"
+camino = "1"
```
camino stuff is re-exported in `saucer` so you can import anything you need from there

EverlastingBugstopper commented on src/command/new/mod.rs:
```diff
@@ -0,0 +1,348 @@
+use camino::Utf8PathBuf;
+use console::Term;
+use dialoguer::Select;
+use saucer::Utf8PathBuf as PathBuf;
```
```suggestion
use saucer::Utf8PathBuf;
```

EverlastingBugstopper commented on src/command/new/mod.rs:
```diff
@@ -0,0 +1,348 @@
+use crate::utils::client::StudioClientConfig;
+use ansi_term::Colour::Cyan;
+use ansi_term::Style;
+use camino::Utf8PathBuf;
```
```suggestion
```

EverlastingBugstopper commented on src/command/new/mod.rs:
```diff
@@ -0,0 +1,348 @@
+  #[clap(flatten)]
+  options: TemplateOpt,
+
+  /// Path to create template at
```
```suggestion
  /// The relative path to create the template directory.
```

EverlastingBugstopper commented on src/command/new/mod.rs:
```diff
@@ -0,0 +1,348 @@
+  pub id: String,
+  pub display: String,
+  pub language: String,
+  pub project_type: String,
```
```suggestion
  pub language: ProjectLanguage,
  pub project_type: ProjectType,
```

these should store the enum types directly instead of converting them to `String`s

EverlastingBugstopper commented on src/command/new/mod.rs:
```diff
@@ -0,0 +1,348 @@
+      return Err(RoverError::new(anyhow!(
+        "You can only create projects in a blank folder. This is to prevent from accidentally overwriting any work."
+      )));
+    }
```
we should remove the `unwrap` here because i think it will cause a panic if the directory doesn't exist. if it doesn't exist, we should create it. 

_something_ like this (i have not tested if it compiles)

```suggestion
    if let Ok(dir_entries) = fs::read_dir(&self.path, "") {
      if dir_entries.count() > 1 {
           return Err(RoverError::new(anyhow!(
              "You can only create projects in a blank folder. This is to prevent from accidentally overwriting any work."
            )));
      }
    } else {
      Fs::create_dir_all(&self.path, "")?;
    }
```

EverlastingBugstopper commented on src/command/new/mod.rs:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    let client = client_config.get_reqwest_client()?;
+    if self.options.template.is_some() {
```
```suggestion
    if let Some(template) = &self.options.template {
```

EverlastingBugstopper commented on src/command/new/mod.rs:
```diff
@@ -0,0 +1,348 @@
+    if self.options.template.is_some() {
+      // User provided template id
+      self.extract_github_tarball(
+        &self.options.template.clone().unwrap().as_str(),
```
```suggestion
        &template,
```

EverlastingBugstopper commented on src/command/new/mod.rs:
```diff
@@ -0,0 +1,348 @@
+        project_type = self.prompt_project_type()?;
+      } else {
+        project_type = ProjectType::from_str(&self.options.project_type.clone().unwrap())?;
+      }
```
something like this would be a bit cleaner

```suggestion
      let project_type = self.options.project_type.unwrap_or_else(|| self.prompt_project_type()?);
```

EverlastingBugstopper commented on src/command/new/mod.rs:
```diff
@@ -0,0 +1,348 @@
+      "{}:\n\t{}",
+      Style::new()
+        .bold()
+        .paint("To learn more GraphQL, head over to our tutorials"),
```
```suggestion
        .paint("To learn more about GraphQL, head over to our tutorials"),
```

EverlastingBugstopper commented on src/command/new/mod.rs:
```diff
@@ -0,0 +1,348 @@
+    saucer::Fs::copy_dir_all(
+      PathBuf::try_from(template_folder_path.join(format!("{}-main", id)).clone())?,
+      PathBuf::try_from(template_folder_path.to_path_buf())?,
+      id,
```
```suggestion
      "",
```

EverlastingBugstopper commented on src/command/new/mod.rs:
```diff
@@ -0,0 +1,348 @@
+    //Delete old unpacked zip
+    saucer::Fs::remove_dir_all(
+      PathBuf::try_from(template_folder_path.join(format!("{}-main", id)).clone())?,
+      id,
```
```suggestion
      "",
```

EverlastingBugstopper commented on src/command/new/mod.rs:
```diff
@@ -0,0 +1,348 @@
+      return Ok(temp_iter.collect());
+    }
+  }
+}
```
```suggestion
    if project_type != ProjectType::ALL && has_language {
      Ok(
        temp_iter
          .filter(|template| {
            project_type.eq(&template.project_type)
              && template.language == self.options.language.clone().unwrap()
          })
          .collect(),
      )
    } else if project_type != ProjectType::ALL {
      Ok(
        temp_iter
          .filter(|template| project_type.eq(&template.project_type))
          .collect(),
      )
    } else if has_language {
      Ok(
        temp_iter
          .filter(|template| template.language == self.options.language.clone().unwrap())
          .collect(),
      )
    } else {
      Ok(temp_iter.collect())
    }
  }
}
```

EverlastingBugstopper reviewed (commented):
looking good! left some comments here mostly just about style, as for structural comments on behavior i left those in the figjam that you linked earlier today. let me know if you would like to pair a bit on this or if my comments are confusing!

EverlastingBugstopper committed:
Merge branch 'main' into avery/alt-subgraph-dev

michael-watson committed:
Support any git url for templates

michael-watson committed:
use saucer instead of camino

michael-watson committed:
Use ProjectLanguage and ProjectType directly

EverlastingBugstopper committed:
chore: updates rover dev versioning strategy (#1288)

EverlastingBugstopper committed:
Merge branch 'main' into avery/alt-subgraph-dev

dbanty commented on .vscode/launch.json:
```diff
@@ -0,0 +1,264 @@
+      },
+      "args": [
+        "new",
+        "/Users/michaelwatson/Desktop/rover-new",
```
Is there a way to template out this file, if it should be committed?

EverlastingBugstopper replied:
i don't think it should be committed

dbanty commented on src/options/template.rs:
```diff
@@ -0,0 +1,149 @@
+
+    /// Type of template project: client or subgraph
+    #[clap(long = "project-type")]
+    #[serde(skip_serializing)]
```
Are these options serialized anywhere in particular, or is `#[serde(skip_serializing)]` just for opting out of analytics? I'd assume we want to find out what the most popular language templates are so we can focus future work.

Maybe we need to manually capture those fields in the interactive mode, though?

EverlastingBugstopper replied:
it's for opting out of analytics

michael-watson committed:
Add docs

EverlastingBugstopper committed:
release: v0.9.0-alpha.8

EverlastingBugstopper committed:
Merge branch 'main' into avery/alt-subgraph-dev

EverlastingBugstopper committed:
fix: harden router spawner/killer (#1293)

This PR fixes #1294 and #1295.

there were some issues on MacOS related to leader/follower socket
addresses that are fixed now, and we also hardened the router killer by
waiting for the health check to start failing before shutting down the
`rover dev` process

EverlastingBugstopper committed:
release: v0.9.0-alpha.9

dbanty committed:
chore: Refactor some `rover new` code (#1290)

Co-authored-by: Dylan Anthony <dbanty@users.noreply.github.com>

EverlastingBugstopper committed:
Merge branch 'avery/alt-subgraph-dev' into watson/rover-new

dbanty committed:
fix: Clarify error message when main `rover dev` session cannot read (#1296)

Co-authored-by: Dylan Anthony <dbanty@users.noreply.github.com>

EverlastingBugstopper committed:
Merge branch 'main' into avery/alt-subgraph-dev

EverlastingBugstopper committed:
fix: remove subgraphs when their SDL cannot be fetched (#1298)

EverlastingBugstopper committed:
release: v0.9.0-alpha.10

dbanty committed:
chore: Refactor `rover new` into `rover template` (#1299)


Co-authored-by: Dylan Anthony <dbanty@users.noreply.github.com>
Co-authored-by: Stephen Barlow <stephen@apollographql.com>

dbanty committed:
Merge branch 'avery/alt-subgraph-dev' into watson/rover-new

dbanty committed:
chore: Remove .vscode/launch.json

dbanty committed:
feat: Add Ariadne template

dbanty committed:
fix: Show error message if there are no templates to list.

dbanty committed:
feat: Add next step calls to action after `template use`

EverlastingBugstopper committed:
feat: require `rover dev` sessions to be the same version (#1314)

fixes #1260 by sending a version check at the start of each `rover dev`
session that ensures the two are the same. this should prevent protocol
bugs if we make changes to it down the line since everything will use
the same version.

EverlastingBugstopper committed:
fix: too many open files error (#1313)

fixes #1311 by updating to the latest `interprocess` commit, we can be
sure we aren't using too many file descriptors per process since they
are cleaned up, even on M1 machines.

EverlastingBugstopper committed:
release: v0.9.0-alpha.11

dbanty committed:
Merge branch 'avery/alt-subgraph-dev' into watson/rover-new

EverlastingBugstopper committed:
feat: adds emojis to rover dev (#1317)

dbanty committed:
chore: Rename Rust template repo to align better with ID naming scheme

EverlastingBugstopper committed:
fix: use channel instead of socket for main `rover dev` session (#1319)

fixes #1315 by using a `crossbeam_channel` to handle all subgraph updates in the main `rover dev` session. this should significantly reduce the overhead of running a single `rover dev` session, and make router shutdowns on ctrl+c much more durable. additionally, if the unix socket is broken, it shouldn't crash the main `rover dev` session, it will log the error and continue gracefully, while the attached session shuts down.

EverlastingBugstopper committed:
Merge branch 'main' into avery/alt-subgraph-dev

EverlastingBugstopper committed:
chore: use --dev mode for router (#1307)

EverlastingBugstopper committed:
release: v0.9.0-rc.0

EverlastingBugstopper committed:
feat: quiet router logs and port/ip assigning (#1321)

router logs are no longer printed by default when running `rover dev`.
they can be printed with `--log info` or `--log debug` and they will be
printed in the same format as `rover` logs are.

additionally this PR renames `--port` to `--supergraph-port` and adds
`--supergraph-address` which allows you to bind the router to a specific
ip.

EverlastingBugstopper committed:
release: v0.9.0-rc.1

EverlastingBugstopper committed:
fix: disallow same URL and small copy updates (#1322)

EverlastingBugstopper committed:
fix: cleans up various error messages and suggestions (#1325)

EverlastingBugstopper committed:
Merge branch 'avery/alt-subgraph-dev' into watson/rover-new

EverlastingBugstopper commented on docs/source/_redirects:
```diff
@@ -3,3 +3,4 @@
 /readmes/ /docs/rover/commands/readmes/
 /graphs/ /docs/rover/commands/graphs/
 /essentials/ /docs/rover/conventions/
+/new/ /docs/rover/commands/new/
```
this should be updated to `/template/`

dbanty replied:
Fixed!

dbanty committed:
docs: rename `new` redirect to `template`

EverlastingBugstopper committed:
cherry pick docs

EverlastingBugstopper committed:
Merge branch 'avery/alt-subgraph-dev' into watson/rover-new

EverlastingBugstopper committed:
feat: prompt for `<PATH>` if it's not specified (#1326)

Co-authored-by: Dylan Anthony <43723790+dbanty@users.noreply.github.com>
Co-authored-by: Dylan Anthony <dbanty@users.noreply.github.com>

EverlastingBugstopper committed:
chore: updates router and adds env var escape hatches (#1329)

EverlastingBugstopper committed:
Merge branch 'avery/alt-subgraph-dev' into watson/rover-new

EverlastingBugstopper committed:
chore: disable java and typescript (for now)

EverlastingBugstopper committed:
Merge branch 'main' into watson/rover-new

EverlastingBugstopper committed:
require tests before release
